### DISCUSSION TITLE
Generic function that returns flatten().fieldErrors fully types

### DISCUSSION THREAD

**Discussion by antoniogiroz** [colinhacks/zod#1735] [Q&A] [answered]
Hello!

I'm trying to write a generic function that return an object with the result of flatten().fieldErrors, but I would like the resulted object will be fully type.

```ts
import { z } from "zod";

function validate<T extends z.ZodTypeAny>(values: z.infer<T>, schema: T) {
    const result = schema.safeParse(values);
    let errors;
    if (!result.success) {
        errors = result.error.flatten().fieldErrors;
    }

    return {
        valid: result.success,
        errors
    }
}

const formData = {
    name: 'Antonio',
    country: 'Spain'
}

const schema = z.object({
    name: z.string(),
})

const { valid, errors } = validate(formData, schema);

console.log(valid);
console.log(errors); // <-- Here I need autocompletion: errors.name
```

I've read https://zod.dev/?id=writing-generic-functions but I'm not able to achieve it.

Here the [Typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJvgNwBQFmArgHYDGMwEdcAbgIYA2wxnMAKYAeACpxBADyF1iAZxQA6AFolRATzCCAgnXUA+ABRduNQXIBcS4HUyCoY-QBo4chgAtBITldEBKRAo4YLgGVjl4KHMabngAXlcPL05FOU47AAVOKDlBYx4zOT9KELhuQXh7HBySkOBMOEMAQii5GJhUmgYGcyLA0tKq6AUE1vbFIahFTG4BGUM-aeBBbmIAUShquVrgtCpSqJgaKDYEIIHgkz4rMdjO7t6nc4vJuWf0Cj2KMLoIrGgQAARAScOAJM6lOicECCKwAcl0MFYLDhT1KYXoMCg6nhAGUwJwbHDPlQfn83J5vGClBAAEYAK0ETEMEJCUJhVmQqSxNgA5gsnmg-KTwvAkFdiC5XuhqRKBHlMADgTBOC4KcliiLfhByopuBB+RLNWSdYI9QbDK9inAAPQ2uDCAC0jrgAAl7II4ABJOB0QSCYhwTg0JFhcDlZisKyvRTswRAA)

Thanks a lot!

**Comment by JacobWeisenburger** [comment, marked as answer]
I think what you need is this:

```ts
errors = result.error.flatten().fieldErrors as {
    [ key in keyof typeof values ]: string[]
}
```

Here's the full example:
```ts
function validate<T extends z.ZodTypeAny> ( values: z.infer<T>, schema: T ) {
    const result = schema.safeParse( values )
    let errors
    if ( !result.success ) {
        errors = result.error.flatten().fieldErrors as {
            [ key in keyof typeof values ]: string[]
        }
    }

    return {
        valid: result.success,
        errors
    }
}

const formData = {
    name: 'Antonio',
    country: 'Spain'
}

const schema = z.object( {
    name: z.string(),
} )

const { valid, errors } = validate( formData, schema )

console.log( valid )
console.log( errors ) // <-- Here I need autocompletion: errors.name
```

**Reply by antoniogiroz** [reply]
Thanks! that's exactly what I wanted
